rtest/include/rtest/action_client_mock.hpp

```diff
@@ -0,0 +1,483 @@
+    {
+      std::lock_guard<std::recursive_mutex> guard(goal_handles_mutex_);
+      auto it = goal_handles_.find(goal_id);
+      if (it != goal_handles_.end()) {
+        goal_handle = it->second.lock();
+      }
+    }
+    if (goal_handle) {
+      stop_callbacks(goal_handle);
+    }
+  }
+
+#if RTEST_ROS_VERSION >= RTEST_ROS_KILTED
+private:
+  std::shared_ptr<void> create_goal_response() const override
+  {
+    using GoalResponse = typename ActionT::Impl::SendGoalService::Response;
+    return std::shared_ptr<void>(new GoalResponse());
+  }
+
+  std::shared_ptr<void> create_result_response() const override
+  {
+    using GoalResultResponse = typename ActionT::Impl::GetResultService::Response;
+    return std::shared_ptr<void>(new GoalResultResponse());
+  }
+
+  std::shared_ptr<void> create_cancel_response() const override
+  {
+    return std::shared_ptr<void>(new CancelResponse());
+  }
+
+  std::shared_ptr<void> create_feedback_message() const override
+  {
+    using FeedbackMessage = typename ActionT::Impl::FeedbackMessage;
+    return std::shared_ptr<void>(new FeedbackMessage());
+  }
+
+  void handle_feedback_message(std::shared_ptr<void> message) override { (void)message; }
+
+  std::shared_ptr<void> create_status_message() const override
+  {
+    using GoalStatusMessage = typename ActionT::Impl::GoalStatusMessage;
+    return std::shared_ptr<void>(new GoalStatusMessage());
+  }
+
+  void handle_status_message(std::shared_ptr<void> message) override { (void)message; }
+#endif
+
+private:
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base_;
+  std::string action_name_;
+  std::map<GoalUUID, typename GoalHandle::WeakPtr> goal_handles_;
+  std::recursive_mutex goal_handles_mutex_;
+};
+}  // namespace rclcpp_action
+
+namespace rtest
+{
+namespace experimental
+{
+template <typename ActionT>
+class ActionClientMock : public MockBase
+{
+public:
+  using Goal = typename ActionT::Goal;
+  using GoalHandle = typename rclcpp_action::ClientGoalHandle<ActionT>;
+  using GoalHandleSharedPtr = typename GoalHandle::SharedPtr;
+  using WrappedResult = typename GoalHandle::WrappedResult;
+  using SendGoalOptions = typename rclcpp_action::Client<ActionT>::SendGoalOptions;
+
+  explicit ActionClientMock(rclcpp_action::ClientBase * client) : client_(client) {}
+  ~ActionClientMock() { StaticMocksRegistry::instance().detachMock(client_); }
+  TEST_TOOLS_SMART_PTR_DEFINITIONS(ActionClientMock<ActionT>)
+
+  MOCK_METHOD(
+    std::shared_future<GoalHandleSharedPtr>,
+    async_send_goal,
+    (const Goal &, const SendGoalOptions &),
+    ());
+  MOCK_METHOD(
+    std::shared_future<WrappedResult>,
+    async_get_result,
+    (typename GoalHandle::SharedPtr),
+    ());
+  MOCK_METHOD(
+    std::shared_future<WrappedResult>,
+    async_cancel_goal,
+    (const GoalHandleSharedPtr &),
+    ());
+  MOCK_METHOD(std::shared_future<WrappedResult>, async_cancel_all_goals, (), ());
+  MOCK_METHOD(bool, action_server_is_ready, (), ());
+
```

Comment:
We are missing the `bool wait_for_action_server(std::chrono::duration)` mock function from `rclcpp_action::ClientBase` (which is not internal API) and from `rclcpp_action::Client`:
- `std::shared_future<CancelResponse::SharedPtr> async_cancel_all_goals(CancelCallback)`
- `void stop_callbacks(GoalHandle::SharedPtr)`
- `void stop_callbacks(const GoalUUID &)`
- `async_cancel_goals_before(const rclcpp::Time &, CancelCallback)`